packages/warning/babel-plugin.js

```diff
@@ -0,0 +1,79 @@
+/**
+ * Internal dependencies
+ */
+const pkg = require( './package.json' );
+
+function babelPlugin( { types: t } ) {
+	const seen = Symbol();
+
+	const typeofProcessExpression = t.binaryExpression(
+		'!==',
+		t.unaryExpression( 'typeof', t.identifier( 'process' ), false ),
+		t.stringLiteral( 'undefined' )
+	);
+
+	const processEnvExpression = t.memberExpression(
+		t.identifier( 'process' ),
+		t.identifier( 'env' ),
+		false
+	);
+
+	const nodeEnvCheckExpression = t.binaryExpression(
+		'!==',
+		t.memberExpression( processEnvExpression, t.identifier( 'NODE_ENV' ), false ),
+		t.stringLiteral( 'production' )
+	);
+
+	const logicalExpression = t.logicalExpression(
+		'&&',
+		t.logicalExpression( '&&', typeofProcessExpression, processEnvExpression ),
+		nodeEnvCheckExpression
+	);
+
+	return {
+		visitor: {
+			ImportDeclaration( path, state ) {
+				const { node } = path;
+				const isThisPackageImport = node.source.value.indexOf( pkg.name ) !== -1;
+
+				if ( ! isThisPackageImport ) {
+					return;
+				}
+
+				const defaultSpecifier = node.specifiers.find(
+					( specifier ) => specifier.type === 'ImportDefaultSpecifier'
+				);
+
+				const { name } = defaultSpecifier.local;
+
+				state.callee = name;
+			},
+			CallExpression( path, state ) {
+				const { node } = path;
+
+				// Ignore if it's already been processed
+				if ( node[ seen ] ) {
+					return;
+				}
```

Comment:
I'm a bit confused on this. Why would this ever happen?

Reply:
I don't fully understand how babel does the transformation. I initially wrote the plugin without this check, but then I get this error when running babel:
```
RangeError: Maximum call stack size exceeded
```
You can reproduce it by commenting that snippet and running `npm run test-unit packages/warning`.

I guess it tries to recursively transform the code. If that's the case, it'll always find the same `warning(...)` call wrapped within the `process.env.NODE_ENV` check.

Reply:
> I guess it tries to recursively transform the code. If that's the case, it'll always find the same `warning(...)` call wrapped within the `process.env.NODE_ENV` check.

Ah, that makes total sense to me in retrospect. Thanks for clarifying.

It would be nice if we could somehow inform the transform traversal to skip that node at the time we're creating it, but it's not immediately clear to me how we'd do that without transforming it in some other way such that it's not picked up by the normal procedure.

I could imagine something in the future where we could tie this into some "opt out" mechanism. An escape hatch something much like [ESLint's `eslint-disable-line`](https://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments). Where the procedure for replacing the node would mark the replacement as skipped.

I think this is fine for now, though.